File: lint-arwaky/src-python/capabilities/naming_rule_checker.py
import re
from typing import Callable
from ..taxonomy import (
    AdapterName,
    ColumnNumber,
    ErrorCode,
    FilePath,
    FilePathList,
    LineNumber,
    LintMessage,
    LintResult,
    LintResultList,
    Severity,
    Count,
    Identity,
    LayerMapVO,
    PatternList,
    LayerNameVO,
)

from ..contract import INamingRuleProtocol, ISourceParserPort

# Regex: snake_case allows leading _, lowercase letters, digits, and underscores
_SNAKE_CASE_RE = re.compile(r"^_?[a-z][a-z0-9]*(_[a-z0-9]+)*$")


class NamingRuleChecker(INamingRuleProtocol):
    """File naming convention checks."""

    @property
    def rule_name(self) -> Identity:
        return Identity(value="naming")

    @staticmethod
    def _has_snake_case(name: str) -> bool:
        """Return True if *name* follows snake_case convention."""
        return bool(_SNAKE_CASE_RE.match(name))

    @staticmethod
    def _resolve_layer_config(
        f: FilePath,
        root_dir: FilePath,
        layer_map: LayerMapVO,
        global_expected: Count,
        global_exceptions: PatternList,
        detect_layer_fn: Callable[[FilePath, FilePath], LayerNameVO | None],
    ):
        """Resolve the expected word count and exception list for *f*.

        Returns (expected, layer_exceptions, layer_name) tuple.
        """
        layer_vo = detect_layer_fn(f, root_dir)
        layer_name = str(layer_vo) if layer_vo else None

        expected = global_expected  # Fixed: global_expected is Count VO with __int__
        layer_exceptions = list(global_exceptions.values)
        if layer_name and layer_name in layer_map:
            layer_cfg = layer_map[layer_name]
            if layer_cfg.word_count is not None:
                expected = layer_cfg.word_count  # This is int, not Count
            if hasattr(layer_cfg, "exceptions") and layer_cfg.exceptions:
                layer_exceptions.extend(layer_cfg.exceptions.values)
        return expected, layer_exceptions, layer_name

    @staticmethod
    def _should_skip_file(basename: str, layer_exceptions: list[str]) -> bool:
        """Return True if *basename* is an exception or a known special file."""
        if basename in layer_exceptions:
            return True
        if basename in ("__init__.py", "main.py", "py.typed"):
            return True
        return False

    def check_file_naming(
        self,
        files: FilePathList,
        root_dir: FilePath,
        layer_map: LayerMapVO,
        global_expected: Count,
        global_exceptions: PatternList,
        results: LintResultList,
        detect_layer_fn: Callable[[FilePath, FilePath], LayerNameVO | None],
    ) -> None:
        """Check that files have the correct number of underscores."""
        for f in files:
            expected, layer_exceptions, layer_name = self._resolve_layer_config(
                f,
                root_dir,
                layer_map,
                global_expected,
                global_exceptions,
                detect_layer_fn,
            )

            basename = str(f).split("/")[-1]
            if self._should_skip_file(basename, layer_exceptions):
                continue

            stem = basename.replace(".py", "")
            # Word count is underscores + 1
            actual_words = stem.count("_") + 1
            if actual_words != int(expected):
                results.append(
                    LintResult(
                        file=f,
                        line=LineNumber(value=1),
                        column=ColumnNumber(value=1),
                        message=LintMessage(
                            value=f"File '{basename}' has {actual_words} words, "
                            f"expected {expected} for layer '{layer_name}'"
                        ),
                        severity=Severity.HIGH,
                        code=ErrorCode(code="NAMING_WORD_COUNT"),
                        source=AdapterName(value="architecture"),
                    )
                )

    def _validate_class_name(
        self,
        f: FilePath,
        cls_info: dict,
        results: LintResultList,
    ) -> None:
        """Append a LintResult if *cls_info* name is not PascalCase."""
        name = cls_info["name"]
        if not name[0].isupper() or "_" in name:
            results.append(
                LintResult(
                    file=f,
                    line=LineNumber(value=cls_info["line"]),
                    column=ColumnNumber(value=cls_info["column"]),
                    message=LintMessage(value=f"Class '{name}' should be PascalCase"),
                    severity=Severity.HIGH,
                    code=ErrorCode(code="NAMING_CLASS_PASCAL_CASE"),
                    source=AdapterName(value="architecture"),
                )
            )

    def check_class_naming(
        self,
        files: FilePathList,
        results: LintResultList,
        source_parser: ISourceParserPort,
    ) -> None:
        """Check that classes follow PascalCase and match file name if single class."""
        for f in files:
            classes = source_parser.get_class_definitions(f)
            for cls_info in classes:
                self._validate_class_name(f, cls_info, results)

    def _validate_function_name(
        self,
        f: FilePath,
        func_info: dict,
        results: LintResultList,
    ) -> None:
        """Append a LintResult if *func_info* name is not snake_case."""
        name = func_info["name"]
        if name.startswith("__") and name.endswith("__"):
            return

        if not self._has_snake_case(name):
            results.append(
                LintResult(
                    file=f,
                    line=LineNumber(value=func_info["line"]),
                    column=ColumnNumber(value=func_info["column"]),
                    message=LintMessage(
                        value=f"Function '{name}' should be snake_case"
                    ),
                    severity=Severity.HIGH,
                    code=ErrorCode(code="NAMING_FUNCTION_SNAKE_CASE"),
                    source=AdapterName(value="architecture"),
                )
            )

    def check_function_naming(
        self,
        files: FilePathList,
        results: LintResultList,
        source_parser: ISourceParserPort,
    ) -> None:
        """Check that functions follow snake_case."""
        for f in files:
            functions = source_parser.get_function_definitions(f)
            for func_info in functions:
                self._validate_function_name(f, func_info, results)

File: lint-arwaky/src-rust/capabilities/naming_rule_checker.rs
// naming_rule_checker — File naming convention checks.
// Implements INamingRuleProtocol: validates snake_case, PascalCase, word counts.

use std::path::Path;
use regex::Regex;
use crate::taxonomy::{
    AdapterName, ColumnNumber, ErrorCode, FilePath, FilePathList,
    LayerNameVO, LintMessage, LintResult, LintResultList, LineNumber,
    Severity, Count, Identity, ArchitectureConfig, ScopeRef, LocationList,
};

pub struct NamingRuleChecker;

impl NamingRuleChecker {
    pub fn new() -> Self {
        Self
    }

    pub fn rule_name(&self) -> Identity {
        Identity::new("naming")
    }

    fn has_snake_case(name: &str) -> bool {
        let re = Regex::new(r"^_?[a-z][a-z0-9]*(_[a-z0-9]+)*$").unwrap();
        re.is_match(name)
    }

    fn get_stem(basename: &str) -> &str {
        if let Some(pos) = basename.rfind('.') {
            &basename[..pos]
        } else {
            basename
        }
    }

    fn is_special_file(basename: &str) -> bool {
        matches!(basename,
            "__init__.py" | "main.py" | "py.typed" | "mod.rs" | "lib.rs"
            | "index.ts" | "index.js" | "app.py"
        )
    }

    fn make_lint_result(file: &str, line: i64, col: i64, code: &str, msg: &str, sev: Severity) -> LintResult {
        LintResult {
            file: FilePath::new(file.to_string()),
            line: LineNumber::new(line),
            column: ColumnNumber::new(col),
            code: ErrorCode::new(code),
            message: LintMessage::new(msg),
            source: AdapterName::new("architecture"),
            severity: sev,
            enclosing_scope: ScopeRef {
                name: "".to_string(),
                kind: "".to_string(),
                file: FilePath::new(""),
                start_line: LineNumber::new(0),
                end_line: LineNumber::new(0),
            },
            related_locations: LocationList::new(Vec::new()),
        }
    }

    /// Check that files have the correct number of underscore-separated words.
    pub fn check_file_naming(
        &self,
        files: &[String],
        root_dir: &str,
        config: &ArchitectureConfig,
        results: &mut Vec<LintResult>,
        detect_layer_fn: impl Fn(&str, &str) -> Option<String>,
    ) {
        let global_expected = config.naming.word_count.value as i32;

        for file in files {
            let basename = Path::new(file)
                .file_name()
                .and_then(|f| f.to_str())
                .unwrap_or("");

            if basename.is_empty() || Self::is_special_file(basename) {
                continue;
            }

            let layer_name = detect_layer_fn(file, root_dir);

            let (expected, exceptions) = if let Some(ref name) = layer_name {
                let key = LayerNameVO::new(name);
                if let Some(def) = config.layers.get(&key) {
                    let exp = if def.word_count.value > 0 {
                        def.word_count.value as i32
                    } else {
                        global_expected
                    };
                    (exp, def.exceptions.values.clone())
                } else {
                    (global_expected, vec![])
                }
            } else {
                (global_expected, vec![])
            };

            if exceptions.contains(&basename.to_string()) {
                continue;
            }

            let stem = Self::get_stem(basename);
            let actual_words = stem.split('_').count() as i32;

            if actual_words != expected {
                let msg = if let Some(ref name) = layer_name {
                    let key = LayerNameVO::new(name);
                    config.layers.get(&key)
                        .filter(|def| !def.word_count_violation_message.value.is_empty())
                        .map(|def| def.word_count_violation_message.value.clone())
                        .unwrap_or_else(|| format!(
                            "AES003 NAMING_CONVENTION: File '{}' has {} words, expected {}.\n\
                            WHY? Strict naming ensures architectural consistency.\n\
                            FIX: Rename to exactly {} words separated by underscores.",
                            basename, actual_words, expected, expected
                        ))
                } else {
                    format!(
                        "AES003 NAMING_CONVENTION: File '{}' has {} words, expected {}.\n\
                        WHY? Strict naming ensures architectural consistency.\n\
                        FIX: Rename to exactly {} words separated by underscores.",
                        basename, actual_words, expected, expected
                    )
                };
                results.push(Self::make_lint_result(file, 1, 1, "AES003", &msg, Severity::HIGH));
            }
        }
    }

    /// Check that class names follow PascalCase.
    pub fn check_class_naming_raw(
        &self,
        file: &str,
        class_name: &str,
        line: i64,
        column: i64,
        results: &mut Vec<LintResult>,
    ) {
        let is_pascal = class_name.chars().next().map(|c| c.is_uppercase()).unwrap_or(false)
            && !class_name.contains('_');

        if !is_pascal {
            results.push(Self::make_lint_result(
                file, line, column,
                "NAMING_CLASS_PASCAL_CASE",
                &format!("Class '{}' should be PascalCase", class_name),
                Severity::HIGH,
            ));
        }
    }

    /// Check that function names follow snake_case (skipping dunder methods).
    pub fn check_function_naming_raw(
        &self,
        file: &str,
        func_name: &str,
        line: i64,
        column: i64,
        results: &mut Vec<LintResult>,
    ) {
        if func_name.starts_with("__") && func_name.ends_with("__") {
            return;
        }

        if !Self::has_snake_case(func_name) {
            results.push(Self::make_lint_result(
                file, line, column,
                "NAMING_FUNCTION_SNAKE_CASE",
                &format!("Function '{}' should be snake_case", func_name),
                Severity::HIGH,
            ));
        }
    }
}
